### DISCUSSION TITLE
Issue with returning Arrays by reference

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#40] [General]
Hi @Theadd ,

I can see you started returning ByRef arrays like [here](https://github.com/Theadd/ds3x/blob/bb06e43f3d9c92aebb7dd86859644823b818b7d9/src/ds3x.accdb.src/modules/ArrayList.cls#L45-L56).

This can become a real issue. If the underlying array gets redimensioned while there are variables pointing to it by reference then you will get a crash. To reproduce, just drop the following code into a standard module:  
```VBA
Option Explicit

Private Const FADF_VARIANT As Long = &H800
Private Const FADF_AUTO As Long = &H1
Private pItems() As Variant

Public Property Get Items() As Variant()
    Static sa As SAFEARRAY_1D
    With sa
        .cDims = 1
        .cbElements = VARIANT_SIZE
        .fFeatures = FADF_VARIANT Or FADF_AUTO
        .cLocks = 10
        .pvData = VarPtr(pItems(0))
        .rgsabound0.cElements = UBound(pItems) - LBound(pItems) + 1
    End With
    MemLongPtr(VarPtrArr(Items)) = VarPtr(sa)
End Property

Sub TestCrash()
    pItems = Array(1, 2, 3)
    
    Dim i() As Variant: i = Items()
    
    Debug.Print i(1) 'Prints 2 as expected
    
    ReDim Preserve pItems(0 To 10)
    
    Debug.Print i(1) 'BOOM (but not always)
End Sub
```
By running ```TestCrash``` you sometime get crashes and you sometime don't depending on what happens with the memory space where the array data used to be (old pvData address).

However, having such a reference returned from a class like you have is even more dangerous because when the class gets destroyed so does the ```Static sa As SAFEARRAY_1D``` gets destroyed and if there is any reference left then it will point to unwanted regions of memory and most certainly cause crashes.

I don't know if there is a workaround for this. I would still like to retain the ability to return arrays by reference as I do it myself [here](https://github.com/cristianbuse/Excel-VBA-Tools/blob/5212e6a08e88541c7419365dc6271520d7986e93/src/ExcelTable.cls#L548-L550). I do have some extra safety like [here](https://github.com/cristianbuse/Excel-VBA-Tools/blob/5212e6a08e88541c7419365dc6271520d7986e93/src/ExcelTable.cls#L145-L164) which basically locks the source array from resizing/deleting but that still doesn't address the issue of having reference to the internal source array after the class has been destroyed.

Any thoughts on how to make this safer?

## Edit 1

I played arround a little more and managed to get this scenario:
- created a new instance of a ```Class1``` in a local variable called ```c```
- added some values to the internal array of ```c```
- returned the internal array by reference and stored it in a local variable called ```v```
- destroyed ```c``` using ```Set c = Nothing```
- created a new ```Class1``` instance using ```Set d = New Class1```
- added other values to the internal array of ```d```
- v now links into the data of instance ```d```

This is very unpredictable behaviour and really dangerous as I would definitely prefer a crash than work with the wrong data.

if we don't find a way to clear all the variables pointing to the array by reference, when the array has been reallocated or destroyed, then we need to stop using this workaround.

**Comment by Theadd** [comment]
Hi @cristianbuse , sorry for late answer. 

I was aware that it was causing instability since it used to crash while debugging in VBE, but I didn't dig that deep. `.Items()` was trying to be the workaround for the bug in the enumerator that you showed me. To prevent it from crashing every now and then in debug mode I changed `.Items()` to a function instead of a property and thus prevent the locals window from automatically calling it to show a preview of its contents. And as it is currently implemented, its only viable for use in a `For Each`. The return value should never be assigned/stored anywhere, but that is not viable in a shared library as I have no way to prevent others from using it the wrong way.

I've been porting the `ArrayList` to twinBASIC, I managed to implement a proper [enumerator](https://github.com/Theadd/ArrayList/blob/main/ArrayListLib/Sources/ArrayList.twin#L125-L129) that doesn't cause that bug and it's also pretty fast, not as fast as `.Items()` but fast enough to be able to get rid of `.Items()`.

As `MemCopy` can't currently be used in twinBASIC and RtlMoveMemory in twinBASIC for Win32 is just as slow as in VBA for Win32 (I target virtual machines w/ 64bit windows and 32bit MS Office 2016 at work, i do it for fun tho) I'm struggling to adapt your idea behind MemCopy to work in twinBASIC but in a reduced version and for a few use cases only, using BSTR and LSet, which I got somehow working already.

By the way, your last two updates in `LibMemory.bas`/`MemCopy` are drastically dropping the performance in, x10 times in Win32 and x5 in Win64 compared to the version you committed in 2023-09-28. Which had that bug on large memory ranges but calling it in chunks avoids hitting the bug, which is also applicable to solve the performance drop in latest versions.

```py
# VBA Win32 - (Move a block of 50000 variants, one by one, 50000 times to the right)

CopyMemory: 15,234
MemCopy (Version: Latest): 40,141

CopyMemory: 17,336
MemCopy (Version: Latest): 35,562

CopyMemory: 15,867
MemCopy (Version: 2023-09-28): 3,598

# VBA Win64 - (Move a block of 50000 variants, one by one, 50000 times to the right) x2 iterations
CopyMemory: 3,426
MemCopy (Version: Latest): 49,988

CopyMemory: 3,684
MemCopy (Version: 2023-09-28): 11,344

# Win64 - MemCopy chunked
ChunkMemCopy (Version: Latest): 11,473
```
 
Splitting calls to MemCopy in chunks
```vb
Public Sub ChunkMemCopy(ByVal DestinationAddress As LongPtr, ByVal SourceAddress As LongPtr, ByVal ByteCount As LongPtr)
#If Win64 Then
    Const chunkSize As Long = VARIANT_SIZE * 2625
#Else
    Const chunkSize As Long = VARIANT_SIZE * 4000
#End If
    If ByteCount <= chunkSize Then
        MemCopy DestinationAddress, SourceAddress, ByteCount
        Exit Sub
    End If
    Dim c As Long, nChunks As Long
    nChunks = CLng((ByteCount - 1) \ chunkSize)
    If DestinationAddress > SourceAddress Then
        MemCopy DestinationAddress + (nChunks * chunkSize), SourceAddress + (nChunks * chunkSize), ByteCount - nChunks * chunkSize
        For c = nChunks - 1 To 0 Step -1
            MemCopy DestinationAddress + (c * chunkSize), SourceAddress + (c * chunkSize), chunkSize
        Next c
    Else
        For c = 0 To nChunks - 1
            MemCopy DestinationAddress + (c * chunkSize), SourceAddress + (c * chunkSize), chunkSize
        Next c
        MemCopy DestinationAddress + (nChunks * chunkSize), SourceAddress + (nChunks * chunkSize), ByteCount - nChunks * chunkSize
    End If
End Sub
```

What I used to get those numbers:
```vb
Private Sub MemCopyInChunks(ByVal Size As Long, ByVal Iterations As Long)
    Dim t() As Variant, i As Long, e As Long
    ReDim t(0 To 2 * Size - 1): For i = 0 To Size - 1: t(i) = i: Next i
    For e = 0 To Iterations - 1
        For i = 0 To Size - 1
            ChunkMemCopy VarPtr(t(i + 1)), VarPtr(t(i)), Size * VARIANT_SIZE
        Next i
        ChunkMemCopy VarPtr(t(0)), VarPtr(t(Size)), Size * VARIANT_SIZE
    Next e
    For i = 0 To Size - 1: Debug.Assert t(i) = i: Next i
End Sub
```

Thanks! =)

**Comment by cristianbuse** [comment]
Thanks @Theadd ,

> I've been porting the ArrayList to twinBASIC, I managed to implement a proper [enumerator](https://github.com/Theadd/ArrayList/blob/main/ArrayListLib/Sources/ArrayList.twin#L125-L129) that doesn't cause that bug and it's also pretty fast, not as fast as .Items() but fast enough to be able to get rid of .Items().

Great. 

BTW, I spent a few more hours yersterday and the ```GetArrayByRef``` from my repo is safe even when reallocation happens and that's because it simply points to the original SAFEARRAY in a wrapped Variant. I totally forgot why I did it that way. However, I still could not figure out how to make it safe for arrays contained by a class instance that goes out of scope. I had a similar problem to solve when I created the ```WeakReference``` repo. There needs to be some kind of reference counting like I did with the ```WeakRefInformer```. If I come up with an idea I will let you know.

>RtlMoveMemory in twinBASIC for Win32 is just as slow as in VBA for Win32

Wow, did not think of that. You probably noticed that yesterday I introduced the ```#If Mac Or TWINBASIC Or (VBA7 = 0) Then``` instead of ```If Mac Then``` which now defaults to ```RtlMoveMemory``` for VBA < 7 and for TB. This had to be done anyway for VBA < 7 but I was so sure TB doesn't have this issue that I did not even test it. Admittedly I don't have TB installed anymore - I was originally contributing to TB (raised tens of issues and even helped with the ```Like``` operator [here](https://github.com/cristianbuse/twinBASIC-Like-Test-Generator)) but then my company confirmed they would never buy TB licenses not because of the pricing but because of the risk of a single person developing it and the enourmous amount of incompatibility issues. Plus, we already have the capability to use other languages like C# to create DLLs.

Anyway, please keep me in the loop and I will hapilly make changes to make this work on TB as well.

> By the way, your last two updates in LibMemory.bas/MemCopy are drastically dropping the performance in, x10 times in Win32 and x5 in Win64 compared to the version you committed in 2023-09-28. Which had that bug on large memory ranges but calling it in chunks avoids hitting the bug, which is also applicable to solve the performance drop in latest versions.

I guess the issue is when using SAFEARRAY and only between size 65536 (old chunk size) up to 16777216 (current max size before defaulting to RtlMoveMemory). I can see you used 63000 for x64 and 64000 for x32 which is pretty much the same thing.

I dropped the chunks because I just did not want to write the extra logic for moving overlapping chunks 😄 . The logic must obviously consider the direction of the copy (left or right) like you did but I figured it would be slightly difficult to make it work within the scope of the same method. I will make an attempt next week to make it work without introducing any new methods i.e. to keep the stack frames low. 

***

Obviously using BSTR is faster but requires having those first 4 bytes with a higher ```Long``` (length) value than the size we want to move. Since I did not want to change the source (just in case it's read-only) the only alternatives I had were:
1) use SAFEARRAY, which I did thus introducing the chunks
2) iterate through the bytes until a suitable value (Long) is found that allows a large block to be moved via BSTR and then copy the remaining with SAFEARRAY. This proved to be very difficult to implement but the biggest issue was that the worst case scenario would be when there is no suitable value and the performance is greatly impacted by iterating many bytes. I might revisit this, there could be value in interating just a few bytes and stop at a certain number - however, I would still need the SAFEARRAY approach as a backup, with chunks.

Just writing this in case this sparks any ideas from you 😄 but mainly useful for me to recap these things . Many thanks!

**Comment by Theadd** [comment]
Hi @cristianbuse ! =)

> You probably noticed that yesterday I introduced the `#If Mac Or TWINBASIC Or (VBA7 = 0) Then` 

Indeed I noticed but it didn't convinced me at all as I was facing very low performance on Win32 =D. Which lead me to dig deeper in the "what and how" you did it in your `MemCopy` from 2023-09-28 to get an adapted version working in tB.

> I can see you used 63000 for x64 and 64000 for x32 which is pretty much the same thing.

I picked them totally based on the `Const chunkSize As Long = 2 ^ 16` line from ur code, although it had to be just less than that since it was an item around that position `(2730 * VARIANT_SIZE aprox, as far as I vaguely recall)` the one causing the bug in the 2023-09-28 version as it was not correctly copied.

I finally got it covered, although, this code is specifically focused on solving the performance gaps that I've been able to find when using RtlMoveMemory in twinBASIC (or a DLL compiled in twinBASIC), being the following:

1.  In **Win64**, moving overlapping memory <u>backwards</u> was **7+ times slower** than in **Win32** or than moving it forward in **Win64**. I don't recall if I tested it on non-overlapping memory.
2. In **Win32**, moving overlapping memory <u>forward</u> was **10+ times slower** than moving it backwards.

Poking around, I found some undocumented methods which seem to be natively ported as they're performing very well. I compared the performance of `VBA.GetMem4`/`VBA.PutMem4` against `MemInt`, being the former ones around 10 times faster in twinBASIC. I've not compared the rest. There was also another method, `VBA.vbaCopyBytes`, which easily solved the _"backwards problem in Win64"_ **(1)**  but not the _"forward"_ one as it doesn't deal with overlapping memory. <sup><em>(Which makes me think that moving overlapped memory backwards probably isn't considered overlapping)</em></sup>

So I only had to adapt the chunked BSTR/LSet approach for the _"move overlapping memory forward"_ **(2)** case.

```vb
    Private DeclareWide PtrSafe Sub RtlMoveMemory Lib "kernel32" (ByVal lpDest As LongPtr, ByVal lpSource As LongPtr, ByVal Size As LongPtr)

    Public Sub MemMoveEx(ByVal Target As LongPtr, ByVal Source As LongPtr, ByVal ByteCount As LongPtr)
        Const MAX_LONG As Long = &H7FFFFFFF&
        If Target > Source Then
            If Source + ByteCount > Target Then
                ' Move Forward + Overlapping
                #If Win64 Then
                    RtlMoveMemory Target, Source, ByteCount
                #Else
                    Const chunkSize As Long = VARIANT_SIZE * 4000
                    Dim c As Long, nChunks As Long = CLng((ByteCount - 1) \ chunkSize)
                    MemMoveBSTR Target + (nChunks * chunkSize), Source + (nChunks * chunkSize), CLng(ByteCount - nChunks * chunkSize)
                    For c = nChunks - 1 To 0 Step -1
                        MemMoveBSTR Target + (c * chunkSize), Source + (c * chunkSize), chunkSize
                    Next c
                #End If
            ElseIf ByteCount < MAX_LONG Then
                ' Move Forward + No overlapping
                VBA.vbaCopyBytes CLng(ByteCount), Target, Source
            Else
                ' Move Forward + No overlapping + More than 2GB exceeding Long
                RtlMoveMemory Target, Source, ByteCount
            End If
        ElseIf ByteCount < MAX_LONG Then
            ' Move Backwards
            VBA.vbaCopyBytes CLng(ByteCount), Target, Source
        Else
            ' Move Backwards + More than 2GB exceeding Long
            RtlMoveMemory Target, Source, ByteCount
        End If
    End Sub
    
    Type UDTStringPointer
        Value As String
    End Type

    Private Sub LSetByVal(ByRef Target As UDTStringPointer, ByVal Value As String)
        LSet Target.Value = Value
    End Sub
    
    ' CopyMemory (Win32/Forward-only) using BSTR/LSet
    Private Sub MemMoveBSTR(ByRef Target As LongPtr, ByRef Source As LongPtr, ByVal ByteCount As Long)
        Dim lbs0 As Long, lbs1 As Long, src As String
        VBA.GetMem4 Source, lbs0
        VBA.GetMem4 Target, lbs1
        VBA.PutMem4 Source, ByteCount - 4
        VBA.PutMem4 Target, ByteCount - 4
        VBA.PutMemPtr VarPtr(src), Source + 4
        LSetByVal Target + 4, src
        VBA.PutMemPtr VarPtr(src), vbNullPtr
        VBA.PutMem4 Target, lbs0
        VBA.PutMem4 Source, lbs0
        If ByteCount > Target - Source Then
            VBA.PutMem4 Target + (Target - Source), lbs1
        End If
    End Sub
```

> Obviously using BSTR is faster but ... Since I did not want to change the source (just in case it's read-only)

By reducing the BSTR approach usage to only those cases in which the destination memory range is overlapping the source memory range, the chances of writing a 4-byte value on a read-only region are... I would like to say "from zero to none", but I don't have enough experience to be able to say that. =)

> make it work within the scope of the same method.

As far as I know, there'd be no problem in that. As you probably have already seen, there are two calls to `MemMoveBSTR` just to remove an extra conditional from inside the loop so, they can be merged back to a single `MemMoveBSTR` that can be written in-place.

But all this just to cover for what twinBASIC could have as "built-in" doesn't convince me either. =\ ='0 =D

Thank again! =)

**Reply by Theadd** [reply]
Allow me to point that it's designed with my specifics in mind, being a `Variant()` array. It doesn't consider cases such as not having enough space for those 4-byte BSTR-size values or having an odd number as ByteCount.

**Comment by cristianbuse** [comment]
Hi @Theadd ,

Did not look into the ```MemCopy``` speed, yet. However, I pushed a [```MemFill```](https://github.com/cristianbuse/VBA-MemoryTools/commit/2a36e6368e23d942c618ec07b39c47d8402544dd) and a [```MemZero```](https://github.com/cristianbuse/VBA-MemoryTools/commit/64016c445c95fa9c42481a491428f691e6a6441a) which are faster than ```FillMemory``` (for VBA7 and up to a certain size of course).

**Comment by cristianbuse** [comment]
Hi @Theadd ,

I pushed a [completely new way to copy memory](https://github.com/cristianbuse/VBA-MemoryTools/commit/af7b9d6d212721ee0e65602e74ec9cbfef468170). Would be grateful if you could run your speed tests on it. For TB it still uses CopyMemory because of [this](https://github.com/twinbasic/twinbasic/issues/1723).

Many thanks!
